sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2706,6 +2707,18 @@ class Dataset[T] private[sql](
 
   /**
    * Persist this Dataset with the default storage level (`MEMORY_AND_DISK`).
+   * @param eager If true, persist the Dataset eagerly.
+   * @group basic
+   * @since 2.3.0
+   */
+  def persist(eager: Boolean): this.type = {
+    persist()
+    if (eager) queryExecution.toRdd.count()
```

Comment:
count() is not efficient though. You can mapPartitions with a no-op function instead.